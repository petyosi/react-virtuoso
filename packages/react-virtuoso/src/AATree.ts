interface NilNode {
  lvl: 0
}

const NIL_NODE: NilNode = { lvl: 0 }

export interface NodeData<T> {
  k: number
  v: T
}

interface NonNilAANode<T> {
  k: number
  v: T
  lvl: number
  l: NonNilAANode<T> | NilNode
  r: NonNilAANode<T> | NilNode
}

export interface Range<T> {
  start: number
  end: number
  value: T
}

export type AANode<T> = NilNode | NonNilAANode<T>

function newAANode<T>(k: number, v: T, lvl: number, l: AANode<T> = NIL_NODE, r: AANode<T> = NIL_NODE): NonNilAANode<T> {
  return { k, v, lvl, l, r }
}

export function empty(node: AANode<any>): node is NilNode {
  return node === NIL_NODE
}

export function newTree<T>(): AANode<T> {
  return NIL_NODE
}

export function remove<T>(node: AANode<T>, key: number): AANode<T> {
  if (empty(node)) return NIL_NODE

  const { k, l, r } = node

  if (key === k) {
    if (empty(l)) {
      return r
    }
    if (empty(r)) {
      return l
    }
    const [lastKey, lastValue] = last(l)
    return adjust(clone(node, { k: lastKey, v: lastValue, l: deleteLast(l) }))
  }
  if (key < k) {
    return adjust(clone(node, { l: remove(l, key) }))
  }
  return adjust(clone(node, { r: remove(r, key) }))
}

export function find<T>(node: AANode<T>, key: number): T | undefined {
  if (empty(node)) {
    return
  }

  if (key === node.k) {
    return node.v
  }
  if (key < node.k) {
    return find(node.l, key)
  }
  return find(node.r, key)
}

export function findMaxKeyValue<T>(node: AANode<T>, value: number, field: 'k' | 'v' = 'k'): [number, T | undefined] {
  if (empty(node)) {
    return [Number.NEGATIVE_INFINITY, undefined]
  }

  if (Number(node[field]) === value) {
    return [node.k, node.v]
  }

  if (Number(node[field]) < value) {
    const r = findMaxKeyValue(node.r, value, field)
    if (r[0] === Number.NEGATIVE_INFINITY) {
      return [node.k, node.v]
    }
    return r
  }

  return findMaxKeyValue(node.l, value, field)
}

export function insert<T>(node: AANode<T>, k: number, v: T): NonNilAANode<T> {
  if (empty(node)) {
    return newAANode(k, v, 1)
  }
  if (k === node.k) {
    return clone(node, { k, v })
  }
  if (k < node.k) {
    return rebalance(clone(node, { l: insert(node.l, k, v) }))
  }
  return rebalance(clone(node, { r: insert(node.r, k, v) }))
}

export function walkWithin<T>(node: AANode<T>, start: number, end: number): NodeData<T>[] {
  if (empty(node)) {
    return []
  }

  const { k, v, l, r } = node
  let result: NodeData<T>[] = []
  if (k > start) {
    result = result.concat(walkWithin(l, start, end))
  }

  if (k >= start && k <= end) {
    result.push({ k, v })
  }

  if (k <= end) {
    result = result.concat(walkWithin(r, start, end))
  }

  return result
}

export function walk<T>(node: AANode<T>): NodeData<T>[] {
  if (empty(node)) {
    return []
  }

  return [...walk(node.l), { k: node.k, v: node.v }, ...walk(node.r)]
}

function last<T>(node: NonNilAANode<T>): [number, T] {
  return empty(node.r) ? [node.k, node.v] : last(node.r)
}

function deleteLast<T>(node: NonNilAANode<T>): AANode<T> {
  return empty(node.r) ? node.l : adjust(clone(node, { r: deleteLast(node.r) }))
}

function clone<T>(node: NonNilAANode<T>, args: Partial<NonNilAANode<T>>): NonNilAANode<T> {
  return newAANode(
    args.k !== undefined ? args.k : node.k,
    args.v !== undefined ? args.v : node.v,
    args.lvl !== undefined ? args.lvl : node.lvl,
    args.l !== undefined ? args.l : node.l,
    args.r !== undefined ? args.r : node.r
  )
}

function isSingle(node: AANode<any>) {
  return empty(node) || node.lvl > node.r.lvl
}

function rebalance<T>(node: NonNilAANode<T>): NonNilAANode<T> {
  return split(skew(node))
}

function adjust<T>(node: NonNilAANode<T>): NonNilAANode<T> {
  const { l, r, lvl } = node
  if (r.lvl >= lvl - 1 && l.lvl >= lvl - 1) {
    return node
  }
  if (lvl > r.lvl + 1) {
    if (isSingle(l)) {
      return skew(clone(node, { lvl: lvl - 1 }))
    }
    if (!empty(l) && !empty(l.r)) {
      return clone(l.r, {
        l: clone(l, { r: l.r.l }),
        r: clone(node, {
          l: l.r.r,
          lvl: lvl - 1,
        }),
        lvl: lvl,
      })
    }
    throw new Error('Unexpected empty nodes')
  }
  if (isSingle(node)) {
    return split(clone(node, { lvl: lvl - 1 }))
  }
  if (!empty(r) && !empty(r.l)) {
    const rl = r.l
    const rlvl = isSingle(rl) ? r.lvl - 1 : r.lvl

    return clone(rl, {
      l: clone(node, {
        r: rl.l,
        lvl: lvl - 1,
      }),
      r: split(clone(r, { l: rl.r, lvl: rlvl })),
      lvl: rl.lvl + 1,
    })
  }
  throw new Error('Unexpected empty nodes')
}

export function keys(node: AANode<any>): number[] {
  if (empty(node)) {
    return []
  }
  return [...keys(node.l), node.k, ...keys(node.r)]
}

export function ranges<T>(node: AANode<T>): Range<T>[] {
  return toRanges(walk(node))
}

export function rangesWithin<T>(node: AANode<T>, startIndex: number, endIndex: number): Range<T>[] {
  if (empty(node)) {
    return []
  }
  const adjustedStart = findMaxKeyValue(node, startIndex)[0]
  return toRanges(walkWithin(node, adjustedStart, endIndex))
}

export function arrayToRanges<T, V>(
  items: T[],
  parser: (item: T) => { index: number; value: V }
): Array<{ start: number; end: number; value: V }> {
  const length = items.length
  if (length === 0) {
    return []
  }

  let { index: start, value } = parser(items[0])

  const result = []

  for (let i = 1; i < length; i++) {
    const { index: nextIndex, value: nextValue } = parser(items[i])
    result.push({ start, end: nextIndex - 1, value })

    start = nextIndex
    value = nextValue
  }

  result.push({ start, end: Number.POSITIVE_INFINITY, value })
  return result
}

function toRanges<T>(nodes: NodeData<T>[]): Range<T>[] {
  return arrayToRanges(nodes, ({ k: index, v: value }) => ({ index, value }))
}

function split<T>(node: NonNilAANode<T>): NonNilAANode<T> {
  const { r, lvl } = node

  return !empty(r) && !empty(r.r) && r.lvl === lvl && r.r.lvl === lvl ? clone(r, { l: clone(node, { r: r.l }), lvl: lvl + 1 }) : node
}

function skew<T>(node: NonNilAANode<T>): NonNilAANode<T> {
  const { l } = node

  return !empty(l) && l.lvl === node.lvl ? clone(l, { r: clone(node, { l: l.r }) }) : node
}
