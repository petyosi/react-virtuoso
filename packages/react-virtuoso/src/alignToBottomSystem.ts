import { domIOSystem } from './domIOSystem'
import { totalListHeightSystem } from './totalListHeightSystem'
import * as u from './urx'

export const alignToBottomSystem = u.system(
  ([{ viewportHeight }, { totalListHeight }]) => {
    const alignToBottom = u.statefulStream(false)

    // keep this for the table component only
    const paddingTopAddition = u.statefulStreamFromEmitter(
      u.pipe(
        u.combineLatest(alignToBottom, viewportHeight, totalListHeight),
        u.filter(([enabled]) => enabled),
        u.map(([, viewportHeight, totalListHeight]) => {
          return Math.max(0, viewportHeight - totalListHeight)
        }),
        u.throttleTime(0),
        u.distinctUntilChanged()
      ),
      0
    )

    return { alignToBottom, paddingTopAddition }
  },
  u.tup(domIOSystem, totalListHeightSystem),
  { singleton: true }
)
