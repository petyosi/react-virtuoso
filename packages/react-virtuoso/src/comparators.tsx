import type { ListRange } from './interfaces'

export function tupleComparator<T1, T2>(prev: [T1, T2] | undefined, current: [T1, T2]) {
  return !!(prev && prev[0] === current[0] && prev[1] === current[1])
}

export function rangeComparator(prev: ListRange | undefined, next: ListRange) {
  return !!(prev && prev.startIndex === next.startIndex && prev.endIndex === next.endIndex)
}
