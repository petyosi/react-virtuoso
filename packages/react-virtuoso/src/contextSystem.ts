import * as u from './urx'

export const contextSystem = u.system(() => {
  const context = u.statefulStream<unknown>(null)

  return {
    context,
  }
})
