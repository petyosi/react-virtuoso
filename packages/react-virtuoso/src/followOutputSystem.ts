import { domIOSystem } from './domIOSystem'
import { initialTopMostItemIndexSystem } from './initialTopMostItemIndexSystem'
import type { FollowOutput, FollowOutputScalarType } from './interfaces'
import { LogLevel, loggerSystem } from './loggerSystem'
import { propsReadySystem } from './propsReadySystem'
import { scrollToIndexSystem } from './scrollToIndexSystem'
import { sizeSystem } from './sizeSystem'
import { stateFlagsSystem } from './stateFlagsSystem'
/* eslint-disable @typescript-eslint/no-unsafe-call */
import * as u from './urx'

function normalizeFollowOutput(follow: FollowOutputScalarType): FollowOutputScalarType {
  if (!follow) {
    return false
  }
  return follow === 'smooth' ? 'smooth' : 'auto'
}

const behaviorFromFollowOutput = (follow: FollowOutput, isAtBottom: boolean) => {
  if (typeof follow === 'function') {
    return normalizeFollowOutput(follow(isAtBottom))
  }
  return isAtBottom && normalizeFollowOutput(follow)
}

export const followOutputSystem = u.system(
  ([
    { totalCount, listRefresh },
    { isAtBottom, atBottomState },
    { scrollToIndex },
    { scrolledToInitialItem },
    { propsReady, didMount },
    { log },
    { scrollingInProgress },
  ]) => {
    const followOutput = u.statefulStream<FollowOutput>(false)
    const autoscrollToBottom = u.stream<true>()
    let pendingScrollHandle: any = null

    function scrollToBottom(followOutputBehavior: FollowOutputScalarType) {
      u.publish(scrollToIndex, {
        index: 'LAST',
        align: 'end',
        behavior: followOutputBehavior,
      })
    }

    u.subscribe(
      u.pipe(
        u.combineLatest(u.pipe(u.duc(totalCount), u.skip(1)), didMount),
        u.withLatestFrom(u.duc(followOutput), isAtBottom, scrolledToInitialItem, scrollingInProgress),
        u.map(([[totalCount, didMount], followOutput, isAtBottom, scrolledToInitialItem, scrollingInProgress]) => {
          let shouldFollow = didMount && scrolledToInitialItem
          let followOutputBehavior: FollowOutputScalarType = 'auto'

          if (shouldFollow) {
            // if scrolling to index is in progress,
            // assume that a previous followOutput response is going
            followOutputBehavior = behaviorFromFollowOutput(followOutput, isAtBottom || scrollingInProgress)
            shouldFollow = shouldFollow && !!followOutputBehavior
          }

          return { totalCount, shouldFollow, followOutputBehavior }
        }),
        u.filter(({ shouldFollow }) => shouldFollow)
      ),
      ({ totalCount, followOutputBehavior }) => {
        if (pendingScrollHandle) {
          pendingScrollHandle()
          pendingScrollHandle = null
        }

        pendingScrollHandle = u.handleNext(listRefresh, () => {
          u.getValue(log)('following output to ', { totalCount }, LogLevel.DEBUG)
          scrollToBottom(followOutputBehavior)
          pendingScrollHandle = null
        })
      }
    )

    function trapNextSizeIncrease(followOutput: boolean) {
      const cancel = u.handleNext(atBottomState, (state) => {
        if (followOutput && !state.atBottom && state.notAtBottomBecause === 'SIZE_INCREASED' && !pendingScrollHandle) {
          u.getValue(log)('scrolling to bottom due to increased size', {}, LogLevel.DEBUG)
          scrollToBottom('auto')
        }
      })
      setTimeout(cancel, 100)
    }

    u.subscribe(
      u.pipe(
        u.combineLatest(u.duc(followOutput), totalCount, propsReady),
        u.filter(([follow, , ready]) => follow && ready),
        u.scan(
          ({ value }, [, next]) => {
            return { refreshed: value === next, value: next }
          },
          { refreshed: false, value: 0 }
        ),
        u.filter(({ refreshed }) => refreshed),
        u.withLatestFrom(followOutput, totalCount)
      ),
      ([, followOutput]) => {
        // activate adjustment only if the initial item is already scrolled to
        if (u.getValue(scrolledToInitialItem)) {
          trapNextSizeIncrease(followOutput !== false)
        }
      }
    )

    u.subscribe(autoscrollToBottom, () => {
      trapNextSizeIncrease(u.getValue(followOutput) !== false)
    })

    u.subscribe(u.combineLatest(u.duc(followOutput), atBottomState), ([followOutput, state]) => {
      if (followOutput && !state.atBottom && state.notAtBottomBecause === 'VIEWPORT_HEIGHT_DECREASING') {
        scrollToBottom('auto')
      }
    })

    return { followOutput, autoscrollToBottom }
  },
  u.tup(sizeSystem, stateFlagsSystem, scrollToIndexSystem, initialTopMostItemIndexSystem, propsReadySystem, loggerSystem, domIOSystem)
)
