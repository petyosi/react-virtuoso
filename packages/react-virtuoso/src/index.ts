export { GroupedVirtuoso, Virtuoso } from './Virtuoso'
export { VirtuosoGrid } from './VirtuosoGrid'
export { TableVirtuoso } from './TableVirtuoso'
export * from './interfaces'
export * from './component-interfaces/Virtuoso'
export * from './component-interfaces/TableVirtuoso'
export * from './component-interfaces/VirtuosoGrid'
export { LogLevel } from './loggerSystem'
export * from './utils/context'
