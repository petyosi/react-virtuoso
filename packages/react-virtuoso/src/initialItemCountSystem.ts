import { initialTopMostItemIndexSystem } from './initialTopMostItemIndexSystem'
import { buildListStateFromItemCount, listStateSystem } from './listStateSystem'
import { propsReadySystem } from './propsReadySystem'
import { sizeSystem } from './sizeSystem'
import * as u from './urx'

export const initialItemCountSystem = u.system(
  ([{ sizes, firstItemIndex, data, gap }, { initialTopMostItemIndex }, { initialItemCount, listState }, { didMount }]) => {
    u.connect(
      u.pipe(
        didMount,
        u.withLatestFrom(initialItemCount),
        u.filter(([, count]) => count !== 0),
        u.withLatestFrom(initialTopMostItemIndex, sizes, firstItemIndex, gap, data),
        u.map(([[, count], initialTopMostItemIndexValue, sizes, firstItemIndex, gap, data = []]) => {
          return buildListStateFromItemCount(count, initialTopMostItemIndexValue, sizes, firstItemIndex, gap, data)
        })
      ),
      listState
    )
    return {}
  },
  u.tup(sizeSystem, initialTopMostItemIndexSystem, listStateSystem, propsReadySystem),
  { singleton: true }
)
