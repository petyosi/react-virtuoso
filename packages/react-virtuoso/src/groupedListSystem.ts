import * as u from './urx'

import { findMaxKeyValue } from './AATree'
import { domIOSystem } from './domIOSystem'
import { hasGroups, sizeSystem } from './sizeSystem'
export interface GroupIndexesAndCount {
  totalCount: number
  groupIndices: number[]
}

export function groupCountsToIndicesAndCount(counts: number[]) {
  return counts.reduce(
    (acc, groupCount) => {
      acc.groupIndices.push(acc.totalCount)
      acc.totalCount += groupCount + 1
      return acc
    },
    {
      totalCount: 0,
      groupIndices: [],
    } as GroupIndexesAndCount
  )
}

export const groupedListSystem = u.system(
  ([{ totalCount, groupIndices, sizes }, { scrollTop, headerHeight }]) => {
    const groupCounts = u.stream<number[]>()
    const topItemsIndexes = u.stream<[number]>()
    const groupIndicesAndCount = u.streamFromEmitter(u.pipe(groupCounts, u.map(groupCountsToIndicesAndCount)))
    u.connect(
      u.pipe(
        groupIndicesAndCount,
        u.map((value) => value.totalCount)
      ),
      totalCount
    )
    u.connect(
      u.pipe(
        groupIndicesAndCount,
        u.map((value) => value.groupIndices)
      ),
      groupIndices
    )

    u.connect(
      u.pipe(
        u.combineLatest(scrollTop, sizes, headerHeight),
        u.filter(([_, sizes]) => hasGroups(sizes)),
        u.map(([scrollTop, state, headerHeight]) => findMaxKeyValue(state.groupOffsetTree, Math.max(scrollTop - headerHeight, 0), 'v')[0]),
        u.distinctUntilChanged(),
        u.map((index) => [index])
      ),
      topItemsIndexes
    )

    return { groupCounts, topItemsIndexes }
  },
  u.tup(sizeSystem, domIOSystem)
)
