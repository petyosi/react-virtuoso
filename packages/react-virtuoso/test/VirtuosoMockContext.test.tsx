import React from 'react'
import { render } from '@testing-library/react'
import { Virtuoso, VirtuosoMockContext, TableVirtuoso, VirtuosoGrid, VirtuosoGridMockContext } from '../src'
import { describe, it, expect } from 'vitest'

describe('VirtuosoMockContext', () => {
  type Item = { id: string; value: string }
  const data: Item[] = [
    { id: '1', value: 'foo' },
    { id: '2', value: 'bar' },
    { id: '3', value: 'baz' },
    { id: '4', value: 'ban' },
    { id: '5', value: 'bam' },
    { id: '6', value: 'baw' },
    { id: '7', value: 'bae' },
    { id: '8', value: 'bat' },
  ]

  describe('List', () => {
    it('correctly renders items', () => {
      const { container } = render(<Virtuoso data={data} />, {
        wrapper: ({ children }) => (
          <VirtuosoMockContext.Provider value={{ viewportHeight: 300, itemHeight: 100 }}>{children}</VirtuosoMockContext.Provider>
        ),
      })

      expect(container).toMatchSnapshot()
    })

    it('correctly renders items with useWindowScroll', () => {
      const { container } = render(<Virtuoso data={data} useWindowScroll />, {
        wrapper: ({ children }) => (
          <VirtuosoMockContext.Provider value={{ viewportHeight: 300, itemHeight: 100 }}>{children}</VirtuosoMockContext.Provider>
        ),
      })

      expect(container).toMatchSnapshot()
    })
  })

  describe('Table', () => {
    it('correctly renders items', () => {
      const { container } = render(<TableVirtuoso data={data} />, {
        wrapper: ({ children }) => (
          <VirtuosoMockContext.Provider value={{ viewportHeight: 300, itemHeight: 100 }}>{children}</VirtuosoMockContext.Provider>
        ),
      })

      expect(container).toMatchSnapshot()
    })

    it('correctly renders items with useWindowScroll', () => {
      const { container } = render(<TableVirtuoso data={data} useWindowScroll />, {
        wrapper: ({ children }) => (
          <VirtuosoMockContext.Provider value={{ viewportHeight: 300, itemHeight: 100 }}>{children}</VirtuosoMockContext.Provider>
        ),
      })

      expect(container).toMatchSnapshot()
    })
  })

  describe('Grid', () => {
    it('correctly renders items', () => {
      const { container } = render(<VirtuosoGrid data={data} />, {
        wrapper: ({ children }) => (
          <VirtuosoGridMockContext.Provider value={{ viewportHeight: 200, viewportWidth: 300, itemHeight: 100, itemWidth: 100 }}>
            {children}
          </VirtuosoGridMockContext.Provider>
        ),
      })

      expect(container).toMatchSnapshot()
    })

    it('renders a single row of items when necessary', () => {
      const { container } = render(<VirtuosoGrid data={data} />, {
        wrapper: ({ children }) => (
          <VirtuosoGridMockContext.Provider value={{ viewportHeight: 100, viewportWidth: 300, itemHeight: 100, itemWidth: 100 }}>
            {children}
          </VirtuosoGridMockContext.Provider>
        ),
      })

      expect(container.querySelectorAll('.virtuoso-grid-item')).toHaveLength(3)
    })

    it('does not overflow the index', () => {
      const fourItemData = data.slice(0, 1)

      const { container } = render(<VirtuosoGrid data={fourItemData} />, {
        wrapper: ({ children }) => (
          <VirtuosoGridMockContext.Provider value={{ viewportHeight: 200, viewportWidth: 300, itemHeight: 100, itemWidth: 100 }}>
            {children}
          </VirtuosoGridMockContext.Provider>
        ),
      })

      expect(container.querySelectorAll('.virtuoso-grid-item')).toHaveLength(1)
    })

    it('correctly renders items with useWindowScroll', () => {
      const { container } = render(<VirtuosoGrid data={data} useWindowScroll />, {
        wrapper: ({ children }) => (
          <VirtuosoGridMockContext.Provider value={{ viewportHeight: 200, viewportWidth: 300, itemHeight: 100, itemWidth: 100 }}>
            {children}
          </VirtuosoGridMockContext.Provider>
        ),
      })

      expect(container).toMatchSnapshot()
    })
  })
})
