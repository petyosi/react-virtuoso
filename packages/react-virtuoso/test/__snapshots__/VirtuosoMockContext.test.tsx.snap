// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VirtuosoMockContext > Grid > correctly renders items 1`] = `
<div>
  <div
    data-testid="virtuoso-scroller"
    data-virtuoso-scroller="true"
    style="height: 100%; outline: none; overflow-y: auto; position: relative;"
    tabindex="0"
  >
    <div
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <div
        class="virtuoso-grid-list"
        data-testid="virtuoso-item-list"
        style="padding-top: 0px; padding-bottom: 100px;"
      >
        <div
          class="virtuoso-grid-item"
          data-index="0"
        >
          Item 0
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="1"
        >
          Item 1
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="2"
        >
          Item 2
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="3"
        >
          Item 3
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="4"
        >
          Item 4
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="5"
        >
          Item 5
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtuosoMockContext > Grid > correctly renders items with useWindowScroll 1`] = `
<div>
  <div
    data-virtuoso-scroller="true"
    style="position: relative; height: 300px;"
  >
    <div
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <div
        class="virtuoso-grid-list"
        data-testid="virtuoso-item-list"
        style="padding-top: 0px; padding-bottom: 100px;"
      >
        <div
          class="virtuoso-grid-item"
          data-index="0"
        >
          Item 0
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="1"
        >
          Item 1
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="2"
        >
          Item 2
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="3"
        >
          Item 3
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="4"
        >
          Item 4
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="5"
        >
          Item 5
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtuosoMockContext > List > correctly renders items 1`] = `
<div>
  <div
    data-testid="virtuoso-scroller"
    data-virtuoso-scroller="true"
    style="height: 100%; outline: none; overflow-y: auto; position: relative;"
    tabindex="0"
  >
    <div
      data-viewport-type="element"
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <div
        data-testid="virtuoso-item-list"
        style="box-sizing: border-box; margin-top: 0px; padding-top: 0px; padding-bottom: 500px;"
      >
        <div
          data-index="0"
          data-item-index="0"
          data-known-size="100"
          style="overflow-anchor: none;"
        >
          Item 0
        </div>
        <div
          data-index="1"
          data-item-index="1"
          data-known-size="100"
          style="overflow-anchor: none;"
        >
          Item 1
        </div>
        <div
          data-index="2"
          data-item-index="2"
          data-known-size="100"
          style="overflow-anchor: none;"
        >
          Item 2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtuosoMockContext > List > correctly renders items with useWindowScroll 1`] = `
<div>
  <div
    data-virtuoso-scroller="true"
    style="position: relative; height: 800px;"
  >
    <div
      data-viewport-type="window"
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <div
        data-testid="virtuoso-item-list"
        style="box-sizing: border-box; margin-top: 0px; padding-top: 0px; padding-bottom: 500px;"
      >
        <div
          data-index="0"
          data-item-index="0"
          data-known-size="100"
          style="overflow-anchor: none;"
        >
          Item 0
        </div>
        <div
          data-index="1"
          data-item-index="1"
          data-known-size="100"
          style="overflow-anchor: none;"
        >
          Item 1
        </div>
        <div
          data-index="2"
          data-item-index="2"
          data-known-size="100"
          style="overflow-anchor: none;"
        >
          Item 2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtuosoMockContext > Table > correctly renders items 1`] = `
<div>
  <div
    data-testid="virtuoso-scroller"
    data-virtuoso-scroller="true"
    style="height: 100%; outline: none; overflow-y: auto; position: relative;"
    tabindex="0"
  >
    <div
      data-viewport-type="element"
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <table
        style="border-spacing: 0; overflow-anchor: none;"
      >
        <tbody
          data-testid="virtuoso-item-list"
        >
          <tr
            data-index="0"
            data-item-index="0"
            data-known-size="100"
            style="overflow-anchor: none;"
          >
            <td>
              Item $
              0
            </td>
          </tr>
          <tr
            data-index="1"
            data-item-index="1"
            data-known-size="100"
            style="overflow-anchor: none;"
          >
            <td>
              Item $
              1
            </td>
          </tr>
          <tr
            data-index="2"
            data-item-index="2"
            data-known-size="100"
            style="overflow-anchor: none;"
          >
            <td>
              Item $
              2
            </td>
          </tr>
          <tr>
            <td
              style="height: 500px; padding: 0px; border: 0px;"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`VirtuosoMockContext > Table > correctly renders items with useWindowScroll 1`] = `
<div>
  <div
    data-virtuoso-scroller="true"
    style="position: relative; height: 800px;"
  >
    <div
      data-viewport-type="window"
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <table
        style="border-spacing: 0; overflow-anchor: none;"
      >
        <tbody
          data-testid="virtuoso-item-list"
        >
          <tr
            data-index="0"
            data-item-index="0"
            data-known-size="100"
            style="overflow-anchor: none;"
          >
            <td>
              Item $
              0
            </td>
          </tr>
          <tr
            data-index="1"
            data-item-index="1"
            data-known-size="100"
            style="overflow-anchor: none;"
          >
            <td>
              Item $
              1
            </td>
          </tr>
          <tr
            data-index="2"
            data-item-index="2"
            data-known-size="100"
            style="overflow-anchor: none;"
          >
            <td>
              Item $
              2
            </td>
          </tr>
          <tr>
            <td
              style="height: 500px; padding: 0px; border: 0px;"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
